docs/DESIGN.md

```diff
@@ -0,0 +1,532 @@
+rate limited. The main downside of this is additional memory if clients 
+often come bearing certificates binding many identities. 
+
+### Authorisation
+
+The server's authorisation subsystem decides which upstreams an 
+authenticated client's connection may be forwarded to.
+
+Let
+- `c` denote an authenticated client
+- `Ids(c)` denote the set of identities associated with a client `c`
+- `Groups(i)` denote the set of client groups that an individual SAN 
+  identity `i` belongs to
+- `UGroups(g)` denote the set of upstream groups that a client group `g` is 
+  authorised to forward to
+- `Upstreams(u)` denote the set of upstreams that are members of the 
+  upstream group `u`
+    
+Let `AuthorisedUpstreams(c)` denote the set of upstreams that an 
+authenticated client `c` is authorised to access.
+
+Then, `AuthorisedUpstreams(c)` is defined as:
+
+```
+AuthorisedUpstreams(c) := union_{i in Ids(c)} (
+                            union_{g in Groups(i)} (
+                                union_{u in UGroups(g)} Upstreams(u) ) )
+```
+
+Clients will be identified for authorisation purposes using Subject Alternative
+Name (SAN) extension to the client's x509 certificate.  This means that
+`Ids(c)` is defined as the set of SAN identities bound to the client's public
+key through the client's x509 certificate.
+
+The server must validate the certification path from a trust anchor to the
+client's certificate during authentication, prior to the authorisation 
+subsystem. Recall that the server has no responsibility to verify these subject
+alternative names, as per
+[rfc5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6)
+that is the job of the CA.
+
+At a minimum, a server must support reading the following kinds of SANs out of
+the client's x509 cert to define `Ids(c)`:
+- email addresses
+- DNS names
+A server may choose to implement support additional kinds of SANs defined in
+[rfc5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6).
+
+Server implementations are encouraged to consider normalisation of SAN data
+when implementing their authorisation subsystems.
+
+A single x509 client certificate for an authenticated client may bind 0, 1 or
+many SANs to the client's public key. The set of upstreams that a client is
+authorised to be forwarded to is defined as the union, over the SANs bound to
+the client's certificate, of the sets of upstreams authorised for each SAN.
+
+In the trivial case where a client certificate binds 0 SANs, then the set of
+upstreams that the client is authorised to be forwarded to will be the empty
+set, i.e., the client will not be authorised to be forwarded to any upstream.
+
+The remaining pieces of input data for authorisation, `Groups(.)`, `UGroups(.)`
+and `Upstreams(.)`, will be sourced by the proof-of-concept server from a 
+JSON configuration file in a path specified by the user.
+
+#### proof-of-concept authorisation data format
```

Comment:
Arguably defining the data format is over specifying it for a poc, but the authorisation data format is part of the user-facing interface of the tool.

I am not the biggest fan of JSON but it is something users will be familiar with and is a somewhat pragmatic choice as there's okay-ish support for JSON in the go standard library (excluding support for JSON schema validation or so on)